programs/quarter_board/main.c

```diff
@@ -0,0 +1,173 @@
+#include "ch.h"
+#include "hal.h"
+
+#include <icu_lld.h>
+#include "chprintf.h"
+#include <inttypes.h>
+
+#include "communication.h"
+
+#define MAX_PULSE_LENGTH 3000
+#define MIN_PULSE_LENGTH 400
+#define NOT_SEEING_VALUE 3000
+#define QUANTITY_OF_SENSORS 2
+
+#define UNUSED(arg) (void)(arg)
+
+int32_t sensors_values[QUANTITY_OF_SENSORS];
+
+thread_t *commanderp;
+thread_t *ir_measurementp;
+
+static void read_pulse(ICUDriver *icup) {
+    uint8_t i;
+    int32_t value;
+
+    if (icup == &ICUD1) {
+        i = 0;
+    } else {
+        i = 1;
+    }
+    
+    value = icuGetWidthX(icup);
+
+    if (value < MAX_PULSE_LENGTH && value > MIN_PULSE_LENGTH) {    
+        sensors_values[i] = value;
+    } else {
+        sensors_values[i] = NOT_SEEING_VALUE;
+    }
+}
+
+void execute_master_command(uint16_t command_id, uint8_t *buff) {
+    switch (command_id) {
+        case 0xAA:  if(buff[0] == 1) led(GREEN, 1);
+                    else led(GREEN, 0);
+
+                    if(buff[1] == 1) led(RED, 1);
+                    else led(RED, 0);
+
+                    break;
+
+        default:    break;
+    }
+
+}
+
+uint8_t *send_data_command(uint16_t command_id, uint8_t data_lenght, uint8_t *buff) {
+    uint32_t values = 0;
+    uint8_t i;
+    switch (command_id) {
+        case 0xFF:  values = sensors_values[0] | (sensors_values[1] << 16);
+                    sensors_values[0] = NOT_SEEING_VALUE;
+                    sensors_values[1] = NOT_SEEING_VALUE;
+                    led(GREEN, 1);
+                    led(RED, 0);
+                    for(i = 0; i < data_lenght; i++) {
+                        buff[i] = ((values >> (4 * i)) & 0xf);
+                    }
+                    break;
+
+        default:    break;           
+    }
+    return buff;
+}
+
+THD_WORKING_AREA(waCommanderThread, 128);
+THD_FUNCTION(CommanderThread, arg) {
+    UNUSED(arg);
+
+    thread_t *serialp;
+    message_data_t *messagep;
+    uint16_t cmd, command_id;
+    uint8_t data_buff[MAX_DATA_CHAR_LENGHT];
+    uint8_t data_length, i, rw_bit;
+
+    while (1) {
+        serialp = chMsgWait();
+        messagep =(message_data_t *) chMsgGet(serialp);
+        
+        cmd = messagep -> operation;
+
+        command_id = (cmd & 0x7ff);
+        rw_bit = (((cmd >> OPERATION_BIT_LENGHT) & 1) + 1);
+        data_length = (cmd >> (OPERATION_BIT_LENGHT + RW_BIT));

```

Comment:
Outermost parentheses are not required here, nor on 2 previous lines. Please remove.